[router] Reduced allocations in router read path

FelixGV opened the pull request:
A lot of multi-key specific state was moved out of VenicePath and into its subclasses, which makes single get queries require less heap space. This includes streaming (chunkedResponse), HAR (helixGroupId and requestId), longTailRetryThresholdMs and responseHeaders.

In addition, all kinds of VenicePaths are also getting their heap size reduced by leveraging shared instances where possible. For example:

- The resourceName, storeName and versionNumber properties are replaced by a single reference to a shared instance of StoreVersionName, which is a new class obtained from a NameRepository (same idea as the PubSubTopicRepository).

- The smartLongTailRetryEnabled, smartLongTailRetryAbortThresholdMs and longTailRetryThresholdMs properties are replaced by a single pointer to a RouterRetryConfig object, which is a simple facade wrapping the VeniceRouterConfig.

- The responseDecompressor is now coming from a map of shared instances in the VenicePathParser.

Miscellaneous:

- Various refactorings enable the VenicePath constructors to have fewer params and to make more of the properties final.

- Made as many properties as possible final in the VeniceRouterConfig.

- Minor tweaks to the main README page.

## How was this PR tested?
<!--
If you're unsure about what to test, where to add tests, or how to run tests, please feel free to ask. We'd be happy to help.
-->
CI.

## Does this PR introduce any user-facing changes?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, choose 'No'.
-->
- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

lusong64 commented on services/venice-router/src/main/java/com/linkedin/venice/router/api/RouterExceptionAndTrackingUtils.java:
```diff
@@ -23,14 +23,25 @@
+    final BiConsumer<AggRouterHttpRequestStats, String> statsRecorder;
+
+    FailureType(BiConsumer<AggRouterHttpRequestStats, String> statsRecorder) {
+      this.reportUnhealthy = statsRecorder == null;
```
So the entire reportUnhealthy is used to recorded whether we have a non-null BiConsumer? Would it possible to just use 

```java
this.reportUnhealthy = Optional.ofNullable(statsRecorder)
    .orElse((stats, store) -> {});
```

FelixGV replied:
The `reportUnhealthy` is not only used within the constructor, it is also used inside the of `metricTracking` function to control some behavior.

Regarding the use of `Optional`, I personally do not find that it improves readability, so not a big fan... but that can be a matter of opinion 🤷

lusong64 replied:
I see.

FelixGV commented:
Carved out a portion of this big-ish PR into this other fairly trivial PR: #1405.

FelixGV commented:
Rebased after merging https://github.com/linkedin/venice/pull/1405 to reduce the PR size, but also added a flaky test fix which still increases it a bit more 😅 ... oh well 🤷

FelixGV commented:
Will be able to further reduce the size of this PR after rebasing on top of #1413.

FelixGV commented:
Rebased

ZacAttack commented on services/venice-router/src/main/java/com/linkedin/venice/router/api/VenicePartitionFinder.java:
```diff
@@ -106,15 +110,17 @@ private VenicePartitioner searchPartitioner(String storeName, int versionNum) {
+        while (candidateVersionIterator.hasNext()) {
+          candidateVersion = candidateVersionIterator.next();
           if (!upToDateVersionsSet.contains(candidateVersion)) {
-            versionByPartitionerMap.remove(candidateVersion);
```
Do we not still need to prune this?

ZacAttack replied:
er wait, nevermind I think?  The candidateVersionIterator is based on this.  So I guess removing it from the iterator has the same effect?

FelixGV replied:
That's right, and from my understanding it is also safer to remove from the iterator...

ZacAttack commented on services/venice-router/src/main/java/com/linkedin/venice/router/api/path/VeniceSingleGetPath.java:
```diff
@@ -32,18 +35,17 @@ public class VeniceSingleGetPath extends VenicePath {
-      String storeName,
-      int versionNumber,
-      String resourceName,
+      StoreVersionName storeVersionName,
```
This is a very satisfying compaction of arguments.

storeName, versionNumber, resourceName -> storeVersionName

FelixGV replied:
One of those cases where doing the efficient thing is also cleaner (more brief, strongly-typed) 😁

ZacAttack commented on services/venice-router/src/main/java/com/linkedin/venice/router/api/path/VenicePath.java:
```diff
@@ -54,69 +47,35 @@ public abstract class VenicePath implements ResourcePath<RouterKey> {
 
   public int getHelixGroupId() {
-    return helixGroupId;
+    return -1;
```
This is kind of neat.  So we're reducing allocations by removing the member variables on the super class?  At the expense of having the child classes implement these?

I'm wondering about some of these though, the requestId one, do child classes need their own version of the requestId member?  Could we just rely on the super and a default function for getRequestId?  Or is there some cost saving that I'm missing?

ZacAttack replied:
Or... maybe we don't actually use request id?

FelixGV replied:
The request ID is only uses for HAR requests I believe.

ZacAttack replied:
I realize the comment line is misplaced.  I mean a few lines up on the getRequestId() function

FelixGV replied:
Yeah I thought so. Both the Helix group ID and request ID are used for HAR, IIUC.

ZacAttack commented on services/venice-router/src/main/java/com/linkedin/venice/router/api/VenicePathParser.java:
```diff
@@ -294,26 +319,6 @@ public VenicePath parseResourceUri(String uri, HTTP_REQUEST request) throws Rout
         }
       }
 
-      boolean decompressOnClient = routerConfig.isDecompressOnClient();
```
How do we dodge needing this?  Is it because this decision is delegated to the responseDecompressor?

FelixGV replied:
This code is moved to a separate function. See line 401.

ZacAttack commented on services/venice-router/src/main/java/com/linkedin/venice/router/api/VeniceResponseDecompressor.java:
```diff
@@ -42,30 +39,20 @@ public class VeniceResponseDecompressor {
 
   public VeniceResponseDecompressor(
-      boolean decompressOnClient,
+      CompressionStrategy clientCompression,
```
Do we even need this?  clientCompression is a private final and I don't see it used outside this classes constructor.

ZacAttack reviewed (approved):
Looks good, dropped a few questions here and there, but I didn't see anything untoward.  Thanks for the clean up!

FelixGV commented:
Thanks for the review!
